airbyte-cdk/python/unit_tests/test/http/test_response_builder.py

```diff
@@ -0,0 +1,150 @@
+# Copyright (c) 2023 Airbyte, Inc., all rights reserved.
+import json
+from copy import deepcopy
+from pathlib import Path
+from typing import Any, Callable, Dict, List, Optional
+from unittest import TestCase
+from unittest.mock import Mock
+
+import pytest
+from airbyte_cdk.test.http.response import HttpResponse
+from airbyte_cdk.test.http.response_builder import HttpResponseBuilder, RecordBuilder, create_builders_from_resource, find_template
+
+_RECORDS_FIELD = "records_field"
+_ID_FIELD = "record_id"
+_CURSOR_FIELD = "record_cursor"
+_ANY_RECORD = {"a_record_field": "a record value"}
+_SOME_RECORDS = {_RECORDS_FIELD: [_ANY_RECORD]}
+_A_RESPONSE_TEMPLATE = _SOME_RECORDS
+
+_RECORD_BUILDER = 0
+_RESPONSE_BUILDER = 1
+
+
+def _record_builder(
+    response_template: Dict[str, Any],
+    records_path: List[str],
+    record_id_path: Optional[List[str]] = None,
+    record_cursor_path: Optional[List[str]] = None,
+) -> HttpResponseBuilder:
+    return create_builders_from_resource(deepcopy(response_template), records_path, record_id_path, record_cursor_path)[0]
+
+
+def _response_builder(
+    response_template: Dict[str, Any],
+    records_path: List[str],
+    pagination_strategy: Optional[Callable[[Dict[str, Any]], None]] = None
+) -> HttpResponseBuilder:
+    return create_builders_from_resource(deepcopy(response_template), records_path, pagination_strategy=pagination_strategy)[1]
+
+
+def _body(response: HttpResponse) -> Dict[str, Any]:
+    return json.loads(response.body)
+
+
+class RecordBuilderTest(TestCase):
+    def test_given_with_id_when_build_then_set_id(self):
+        builder = _record_builder({_RECORDS_FIELD: [{_ID_FIELD: "an id"}]}, [_RECORDS_FIELD], [_ID_FIELD])
+        record = builder.with_id("another id").build()
+        assert record[_ID_FIELD] == "another id"
+
+    def test_given_nested_id_when_build_then_set_id(self):
```

Comment:
It seems like this is for the ID within the record but it makes sense that it would apply for the ID in the HTTP response as well. I'll update that

Reply:
One other thing I don't like it the interface for composite paths needs to have the same order as how the keys are defined. For example, if we have `CompositePath([FieldPath("first"), FieldPath("second")])`, we need to have `with_id(["value for first", "value for second"])` and it would be easy to do the mistake `with_id(["value for second", "value for first"])`

Reply:
> do we actually need composite key if this is only a pointer to a path in the HTTP response?

I'm fine with removing this. I added it because the [link you provided](https://github.com/airbytehq/airbyte/blob/master/airbyte-cdk/python/airbyte_cdk/sources/streams/core.py#L277-L278) indicated that this is a concept that exists. However, we don't need this for stripe. Even more, we only need to set the ID for pagination purposes so an API having composite keys might not even use `with_id`

> I think a record with a composite primary key can just be a record with at least fields

Can you expand on that by providing an example? I would like to make sure I understand what you mean before reverting

Reply:
from the connector's perspective, the primary key is just a sort of label for the platform+destination. so long as the records have the non-field from the composite key, it's valid and it doesn't matter:

{
  "id": "1234",
  "region": "us"
}

[id, region] can be a pk but it doesn't matter to the http response which might look like:
{
"data": [{"id": "1234", "region": "us"}]
}

Reply:
After a sync, we agreed that we don't know enough today and we don't have a clear use case for CompositePath. Let's push this to when we will have a case and hence more information to take the right decision